src/domain/task-system/services/src/task_executor_impl.rs

```diff
@@ -77,20 +72,32 @@ impl TaskExecutorImpl {
     }
 }
 
-/////////////////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////
 
 #[async_trait::async_trait]
 impl TaskExecutor for TaskExecutorImpl {
     // TODO: Error and panic handling strategy
     async fn run(&self) -> Result<(), InternalError> {
         loop {
-            let task_id = self.task_sched.take().await.int_err()?;
-            let mut task = Task::load(task_id, self.event_store.as_ref())
-                .await
-                .int_err()?;
+            let mut task = DatabaseTransactionRunner::run_transactional(
```

Comment:
Currently you are opening and closing transaction when loading a task and when saving a task, but what if task execution itself requires DB access?

I would expect the `run` loop to look more like this:
```rust
async fn run(&self) {
  loop {
    // Scheduler is a special **singleton** service - it cannot inject DB transactions and
    // thus should manage transactions on its own.
    let task_id = self.scheduler.take();

    run_transactional(|...| {
      let event_store = ...;
      let task = Task::load(event_store);

      // Run the task under the transaction
      Self::execute_task(task);

      // Store task state at the end of the transaction
      task.update(...);
      task.finish(...);
      task.save(event_store);
    });
  }
}
```
This way, assuming all task actions are transactional - we would get full atomicity.

Reply:
The only tricky point I see are `publish_task_running` /  `publish_task_finished`. Our event bus is not yet transactional, so having them inside `run_transactional` would work. But once we make event bus use DB and transaction we will want a task to be marked "running" to become visible to all other users - thus be committed.

Let's ignore this error for now, and in future perhaps we will make `publish_task_running` a part of `scheduler.take()` - so that the task is marked "running" at the same time as it's associated with the specific executor.

Reply:
I guess another way to look at this problem is that a task may involve long-running data processing and we wouldn't like to hold the DB transaction open for the entire duration of the task.

Our task system may require separating the stages like:
- prepare - read-only DB access to gather info for the execution into a plan)
- execute - performing planned long-running work
- commit - applying all the updates to the DB

But for the time being holding a transaction open should not be too problematic.

I don't really see the alternative, as once we progressively start adding DB dependencies (e.g. managing dataset metadata and references in Postgres) - tasks will need transactional context.

What do you think @zaychenko-sergei ?

Reply:
I think that execution of the tasks should clearly be outside the scope of the transactions that prepare and commit. We should assume that task execution is a heavyweight computational task, potentially running on other computer (future operator). If there will be another type of task that requires it's own transactions for different data, it would have to control it by itself, but I don't think that would be related to tasks management - tasks should be in saved state before that starts.

Reply:
> Our event bus is not yet transactional, so having them inside run_transactional would work

Event bus will not be transactional. By definition, it's an asynchronous notification of an already saved state change. Rollback of the main change that caused the event should not be possible by the time the notification starts. So all event issuing should happen outside transactions. The handlers on their own may open new transactions to react to the change.

Reply:
> Event bus will not be transactional

We previously discussed two-tier event bus model:
- intra-process one, for decoupling components - this one carries events in memory and all subscribers can react to them in the scope of the original transaction
- inter-process one, for communication between services and instances of the same service - publishing events to this one can be made transactional, but instances react to them in independent transactions (eventually consistent)